BCS.py

```diff
@@ -0,0 +1,104 @@
+import numpy as np
+
+def consistent(M, b):
+    """
+    Check if the binary system Mx = b (mod 2) is consistent
+    """
+    aq = list(range(M.shape[0]))
+    A = M.copy()
+    b1 = b.copy()
+    m, n = A.shape
+    row = 0
+
+    for col in range(n):
+        if row >= m:
+            break
+        # Find a pivot row in the set defined by aq.
+        pivot = None
+        for r in range(row, m):
+            if A[aq[r], col] == 1:
+                pivot = r
+                break
+        if pivot is None:
+            continue 
+        # Swap the pivot row into the current position.
+        if pivot != row:
+            A[[aq[row], aq[pivot]]] = A[[aq[pivot], aq[row]]]
+            b1[[aq[row], aq[pivot]]] = b1[[aq[pivot], aq[row]]]
+        # Eliminate the 1's below the pivot in this column.
+        for r in range(row + 1, m):
+            if A[aq[r], col] == 1:
+                A[aq[r]] = (A[aq[r]] + A[aq[row]]) % 2
+                b1[aq[r]] = (b1[aq[r]] + b1[aq[row]]) % 2
+        row += 1
+    for r in range(m):
+        if np.all(A[aq[r]] == 0) and b1[aq[r]] == 1:
+            return False
+    return True
+
+class NonlocalGame:
+    def __init__(self, prob: np.ndarray, pred: np.ndarray, reps: int):
+        self.prob = prob
+        self.pred = pred
+        self.reps = reps
+
+    def commuting_measurement_value_upper_bound(self, k) -> float:
+        if hasattr(self, '_perfect'):
+            return 1.0 if self._perfect else 0
+        return 0
+
+    def has_perfect_commuting_measurement_strategy(self, tol: float = 1e-6) -> bool:
+        comm_value = self.commuting_measurement_value_upper_bound(k=2)
+        return np.isclose(comm_value, 1.0, atol=tol)
```

Comment:
This doesn't seem to be doing what we want. We can already compute the commuting measurement value numerically. We want a method to let us know if a perfect strategy exists or not. In doing so, we specifically do not want to invoke the upper bound function, we instead want to use the group theoretic approach that is outlined in the issue.